hazelcast/src/hazelcast/client/impl/statistics/Statistics.cpp

```diff
@@ -0,0 +1,234 @@
+                const std::string Statistics::NEAR_CACHE_CATEGORY_PREFIX("nc.");
+                const std::string Statistics::FEATURE_SUPPORTED_SINCE_VERSION_STRING("3.9");
+                const int Statistics::FEATURE_SUPPORTED_SINCE_VERSION = impl::BuildInfo::calculateVersion(
+                        FEATURE_SUPPORTED_SINCE_VERSION_STRING);
+
+                Statistics::Statistics(spi::ClientContext &clientContext) : clientContext(clientContext),
+                                                                            clientProperties(
+                                                                                    clientContext.getClientProperties()),
+                                                                            logger(clientContext.getLogger()),
+                                                                            periodicStats(*this) {
+                    this->enabled = clientProperties.getStatisticsEnabled().getBoolean();
+                }
+
+                void Statistics::start() {
+                    if (!enabled) {
+                        return;
+                    }
+
+                    int64_t periodSeconds = clientProperties.getStatisticsPeriodSeconds().getLong();
+                    if (periodSeconds <= 0) {
+
+                        int64_t defaultValue = util::IOUtil::to_value<int64_t>(
+                                clientProperties.getStatisticsPeriodSeconds().getDefaultValue());
+                        logger.warning() << "Provided client statistics "
+                                         << clientProperties.getStatisticsPeriodSeconds().getName()
+                                         << " can not be less than or equal to 0. You provided " << periodSeconds
+                                         << " seconds as the configuration. Client will use the default value of "
+                                         << defaultValue << " instead.";
+                        periodSeconds = defaultValue;
+                    }
+
+                    schedulePeriodicStatisticsSendTask(periodSeconds);
+
+                    logger.info() << "Client statistics is enabled with period " << periodSeconds << " seconds.";
+
+                }
+
+                void Statistics::schedulePeriodicStatisticsSendTask(int64_t periodSeconds) {
+                    clientContext.getClientExecutionService().scheduleWithRepetition(
+                            boost::shared_ptr<util::Runnable>(new CollectStatisticsTask(*this)), 0,
+                            periodSeconds * 1000);
+                }
+
+                boost::shared_ptr<connection::Connection> Statistics::getOwnerConnection() {
+                    connection::ClientConnectionManagerImpl &connectionManager = clientContext.getConnectionManager();
+                    boost::shared_ptr<connection::Connection> connection = connectionManager.getOwnerConnection();
+                    if (NULL == connection.get()) {
```

Comment:
Preference. Once, i was told that this is a better practice to make it more observable to the reader of the code, since then i use this practice.

Reply:
I think in CPP it is useful to prevent assignments in if statement.

For example `connection.get() = NULL` wont result in compilation error because of the missing `=` and it can create bugs that are hard to debug which makes sense to write `NULL == connection.get()` as if you write `NULL = connection.get()` it will result in compilation error. I think it is specific to CPP. In Java `connection.get() = NULL` will also result in compilation error.